src/pipeline/cli_pipeline.c

```diff
@@ -0,0 +1,703 @@
+	d_list_t		  *head = NULL;
+
+	p_auxi                          = tse_task_stack_push(api_task, sizeof(*p_auxi));
+	p_auxi->opc                     = DAOS_PIPELINE_RPC_RUN;
+	p_auxi->map_ver_req             = map_ver;
+	p_auxi->omd_id                  = obj_md->omd_id;
+	p_auxi->api_task                = api_task;
+	head                            = &p_auxi->shard_task_head;
+	D_INIT_LIST_HEAD(head);
+
+	*pipeline_auxi = p_auxi;
+}
+
+int
+dc_pipeline_run(tse_task_t *api_task)
+{
+	daos_pipeline_run_t          *api_args = dc_task_get_args(api_task);
+	struct pl_obj_layout         *layout   = NULL;
+	daos_handle_t                 coh;
+	struct daos_obj_md            obj_md;
+	daos_handle_t                 poh;
+	struct dc_pool               *pool;
+	struct daos_oclass_attr      *oca;
+	int                           rc;
+	d_list_t                      *shard_task_head = NULL;
+	daos_unit_oid_t               oid;
+	uint32_t                      map_ver = 0;
+	uuid_t                        coh_uuid;
+	uuid_t                        cont_uuid;
+	struct pipeline_auxi_args    *pipeline_auxi;
+	uint32_t                      nr_grps;
+	struct shard_task_sched_args  sched_arg;
+	int                           total_shards;
+	int                           total_replicas;
+	int                           shard;
+	struct pipeline_comp_cb_args  comp_cb_args;
+
+	coh = dc_obj_hdl2cont_hdl(api_args->oh);
+	rc  = dc_obj_hdl2obj_md(api_args->oh, &obj_md);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+	poh  = dc_cont_hdl2pool_hdl(coh);
+	pool = dc_hdl2pool(poh);
+	if (pool == NULL) {
+		D_WARN("Cannot find valid pool\n");
+		D_GOTO(out, rc = -DER_NO_HDL);
+	}
+	obj_md.omd_ver = dc_pool_get_version(pool);
+
+	/** get layout */
+
+	rc             = pipeline_create_layout(coh, pool, &obj_md, &layout);
+	dc_pool_put(pool);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+	rc = dc_cont_hdl2uuid(coh, &coh_uuid, &cont_uuid);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+	/** object class */
+
+	oca = daos_oclass_attr_find(obj_md.omd_id, &nr_grps);
+	if (oca == NULL) {
+		D_DEBUG(DB_PL, "Failed to find oclass attr\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+	if (daos_oclass_is_ec(oca)) {
+		D_DEBUG(DB_PL, "EC objects not supported for pipelines\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	/** Not supporting EC objects yet, so #groups x #replicas = #shards */
+
+	total_replicas = layout->ol_grp_size;
+	total_shards   = layout->ol_grp_nr * total_replicas;
+
+	/**
+	 * Pipelines are read only for now, no transactions needed.
+	 * Ignoring api_args->th for now...
+	 */
+
+	if (map_ver == 0)
+		map_ver = layout->ol_ver;
+
+	/** -- Register completion call back function for full operation */
+
+	comp_cb_args.api_args       = api_args;
+	comp_cb_args.oca            = oca;
+	comp_cb_args.total_shards   = total_shards;
+	comp_cb_args.total_replicas = total_replicas;
+
+	pipeline_create_auxi(api_task, map_ver, &obj_md, &pipeline_auxi);
+
+	rc = tse_task_register_comp_cb(api_task, pipeline_comp_cb, &comp_cb_args,
```

Comment:
If there is always single shard task in dc_pipe_line(), then probably we do not need create another shard task here?  or  it will eventually have multiple shards task here?

Reply:
@wangdi1 I imagine that when EC objects get supported, it may  need to have multiple tasks. But maybe, even in that case, this is still an overkill (for regular objects). So yeah, not sure what the best course of action is. It works now, but maybe it is better to just call the function directly instead of creating a task there.